feat: creating a new test suite for running snippet datasets within a VM (#422)
readmeio/api: .github/CONTRIBUTING.md

> # Contributing

Author: @kanadgupta You should review this contributing guide I'm adding to the repo.

<img align="right" width="25%" style="margin-bottom: 2em" src="https://owlbert.io/images/owlberts-png/camp.png.png">

## Commit Conventions

For our general commit conventions please consult our organization contributing guidelines [here](https://github.com/readmeio/.github/blob/main/.github/CONTRIBUTING.md).

## Help

If you need help don't be afraid to open up an issue!

## Packages

> ### `httpsnippet-client-api` snippet

Reviewer: Wanna run this file through your VS Code formatter? I think there should be a newline between these headers

If you wish to add or change functionality to how `api` snippets are generated, your changes should go within the `httpsnippet-client-api` package in `packages/`. We have a nifty test suite set up there that'll run pre-made JS code snippets for the [api](https://npm.im/api) package within a virtual environment, with mocked out HTTP requests, to ensure that everything is well and good. It's pretty neat!

To add a new test into this suite you should do the following:

1. Create a new directory for your test in `packages/httpsnippet-client-api/test/__datasets__`. This directory should be a short and clear description of what you're testing.

>     * If you need to give additional context to your test please do add a `README` in this directory!

Reviewer: ```suggestion
    * If you need to give additional context to your test, please add a `README` in this directory!
```

2. In this new directory you'll need to create 4 files: `har.json`, `openapi.json`, `mock.json`, and `output.js`.

> #### har.json

Reviewer: ```suggestion
#### `har.json`
```

> The way that `httpsnippet-client-api` works is that it takes a HAR object and constructs a snippet according to it, and what the matching operation in an OpenAPI definition, say the API request should function.

Reviewer: This reads a bit weirdly so I tried rewriting it, but feel free to mess with it further:
```suggestion
`httpsnippet-client-api` works by taking a HAR object and its matching operation in an OpenAPI definition and constructing a code snippet based upon them.
```

This file should contain a fully defined [HAR object](http://www.softwareishard.com/blog/har-12-spec/) for your API request.

If you need to use dynamic data (like base64 encoding an API key) then this file can be renamed to `har.js` and the HAR can be exported with `module.exports`.

> #### openapi.json

Reviewer: ```suggestion
#### `openapi.json`
```

> Within `openapi.json` you'll want to spec out the OpenAPI definition that will be powering this request. This definition should be fully valid to the spec.

Reviewer: Idea I just had—since we require that every spec is valid, would it make sense to have a generic `beforeEach` test that asserts this using our validator?

> If you'd like to reference one of the definitions that we provide in the [@readme/oas-examples](https://npm.im/@readme/oas-examples) package then you can rename `openapi.json` to `openapi.js` and have a single line of the following and the test suite will pick it up!

Reviewer: I noticed that you did this and this is really nice! What are your thoughts on us doing this kind of thing in our main codebase? I noticed a PR the other day where you were manually incorporating updates from the `oas-examples` repo and this pattern might help us avoid that?

Reviewer: ```suggestion
If you'd like to reference one of the definitions that we provide in the [`@readme/oas-examples`](https://npm.im/@readme/oas-examples) package then you can rename `openapi.json` to `openapi.js` and have a single line of the following and the test suite will pick it up!
```

Author: I definitely want to do that but I need to figure out a way to incorporate that with the `packages/testing-utils` package compilation process and how https://preview.readme.io uses that.

```js
module.exports = require('@readme/oas-examples/path/to/spec');
````

> Additionally if you need help verifying that your OpenAPI definition is valid you can run `npx rdme validate path/to/your/definition.json`.

Reviewer: 😎

Reviewer: this also kinda ties into my comment above about validating specs in our tests

Author: Added pre-test validation!

> #### mock.json

Reviewer: ```suggestion
#### `mock.json`
```

> This file is what is going to tell our test suite what endpoints your API test will be accessing so we can mock out those requests for you. Behind the scenes we use the amazing [nock](https://npm.im/nock) library for this. Generating this data is a bit difficult, but you can use `nock` itself to guide you along.

Reviewer: ```suggestion
This file is what is going to tell our test suite what endpoints your API test will be accessing so we can mock out those requests for you. Behind the scenes we use the amazing [`nock`](https://npm.im/nock) library for this. Generating this data is a bit difficult, but you can use `nock` itself to guide you along.
```

Say you have the following snippet:

```js
const sdk = require('api')('https://raw.githubusercontent.com/readmeio/api/main/packages/httpsnippet-client-api/test/__datasets__/query/openapi.json');

sdk.get('/anything', {foo: ['bar', 'baz'], baz: 'abc', key: 'value'})
  .then(res => console.log(res))
  .catch(err => console.error(err));
```

Running this normally you'll receive a JSON object back from the https://httpbin.org server referenced in that OpenAPI definition:

```js
{
  args: { baz: 'abc', foo: [ 'bar', 'baz' ], key: 'value' },
  data: '',
  files: {},
  form: {},
  headers: {
    Accept: '*/*',
    'Accept-Encoding': 'gzip,deflate',
    Host: 'httpbin.org',
    'User-Agent': 'api (node)/4.2.0',

>     'X-Amzn-Trace-Id': 'REDACTED'

Reviewer: weird lol, is this an httpbin thing?

Author: Yeah httpbin is served from AWS I guess and adds this into the headers that come back in its responses. Probably from a gateway sitting in front of their servers?

  },
  json: null,
  method: 'GET',
  origin: 'REDACTED',
  url: 'https://httpbin.org/anything?key=value&foo=bar&foo=baz&baz=abc'
}
```

To have `nock` interpret this request into data that we can use to mock out requests back with `nock` you can add modify your snippet as such:

```js
const nock = require('nock');

nock.recorder.rec({

>   output_objects: true,

Reviewer: Interesting, I've never used this `output_objects` option before! I really dig this approach, makes writing tests a loooot quicker

    "accept-encoding": [
      "gzip,deflate"
    ],
    "connection": [
      "close"
    ],
    "host": "httpbin.org"
  },
  "responseIsBinary": false
}
<<<<<<-- cut here -->>>>>>

{
  args: { baz: 'abc', foo: [ 'bar', 'baz' ], key: 'value' },
  data: '',
  files: {},
  form: {},
  headers: {
    Accept: '*/*',
    'Accept-Encoding': 'gzip,deflate',
    Host: 'httpbin.org',
    'User-Agent': 'api (node)/4.2.0',
    'X-Amzn-Trace-Id': 'REDACTED'
  },
  json: null,
  method: 'GET',
  origin: 'REDACTED',
  url: 'https://httpbin.org/anything?key=value&foo=bar&foo=baz&baz=abc'
}
```

Now all you have to do is take the data within the `<<<<<<-- cut here -->>>>>>` block and slim it down to only what you need for your API mock (you can always skip `response`):

```json
{
  "scope": "https://httpbin.org:443",
  "method": "GET",
  "path": "/anything?key=value&foo=bar&foo=baz&baz=abc",
  "status": 200
}
```

Some things you should be aware of:

* You can leave `response` out of `mock.json` as our test suite automatically mocks out a `response` to use so we can assert that what we mocked as a response was what we got.
* If your API doesn't have a request body you can leave `body` off.

> * If your API request doesn't use any unique headers (eg. `Authentication` or `Content-Type`), then you can generally leave `reqheaders` off.

Reviewer: ```suggestion
* If your API request doesn't use any unique headers (e.g. `Authentication` or `Content-Type`), then you can generally leave `reqheaders` off.
```

* `status` should always be 200.
* If you're testing something like `multpart/form-data` requests and need to regex a boundary you can rename `mock.json` to `mock.js` and add your regex in. See the `multipart-*` test for examples of this.

> #### output.js

Reviewer: ```suggestion
#### `output.js`
```

Now that you've done all the hard work for setting up your unit test all that's left is to craft the snippet you're expecting we should generate for your use case. You have a couple options for this:

1. Write it manually.
2. Run the test suite without one so it fails, and then populate what we thought it should be from the data in `har.json` and `openapi.json` into `output.js`.

> After that, and all your tests are passing, then you're good! Thanks for helping out!

Reviewer: This is super thorough and I learned a ton by just reading this—thanks for doing this!
